chainstate/src/detail/mod.rs

```diff
@@ -408,56 +399,174 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
+                let chainstate_ref = self.make_db_tx_ro().map_err(BlockError::from).log_err()?;
+                let saved_block_index =
+                    get_existing_block_index(&chainstate_ref, &block_id).log_err()?;
+
+                assert!(saved_block_index.status().is_ok());
+                return Ok(reorg_occurred.then_some(saved_block_index));
+            }
+            Err(BlockIntegrationError::BlockCommitError(block_id, attempts_count, db_err)) => {
+                return Err(BlockError::DbCommitError(
+                    attempts_count,
+                    db_err,
+                    DbCommittingContext::Block(block_id),
+                ))
+            }
+            Err(BlockIntegrationError::OtherDbError(db_err)) => {
+                return Err(BlockError::StorageError(db_err));
+            }
+            Err(BlockIntegrationError::ConnectBlockErrorDuringReorg(
+                err,
+                status,
+                first_invalid_block_id,
+            )) => (err, status, Some(first_invalid_block_id)),
+            Err(BlockIntegrationError::OtherValidationError(err, status)) => (err, status, None),
+        };
+
+        // Update the block status; note that this is needed even if we're going to call
+        // invalidate_side_chain_block below, because it expects that all block indices
+        // already exist (also, it will update this block's status, indicating that it has
+        // a bad parent).
+        {
+            let block_index = block_index.with_status(status);
+            // Note: we already have an error to return, so we ignore the result of
+            // the following call.
+            let _result = self
+                .with_rw_tx(
+                    |chainstate_ref| chainstate_ref.set_block_status(&block_index),
+                    |attempt_number| {
+                        log::info!(
+                            "Updating status for block {block_id}, attempt #{attempt_number}"
+                        );
+                    },
+                    |attempts_count, db_err| {
+                        BlockError::DbCommitError(
+                            attempts_count,
+                            db_err,
+                            DbCommittingContext::BlockStatus(block_id),
+                        )
+                    },
+                )
+                .log_err();
         }
 
-        let block_index = block_index.with_status(block_status);
+        if let Some(first_invalid_block_id) = first_invalid_block_id {
+            // Again, we ignore the result here.
+            let _result = self.invalidate_side_chain_block(&first_invalid_block_id).log_err();
+        }
+
+        Err(err)
+    }
 
-        // Update block index status.
-        let status_update_result = self.with_rw_tx(
-            |chainstate_ref| chainstate_ref.set_block_status(&block_index),
+    fn invalidate_side_chain_block(&mut self, block_id: &Id<Block>) -> Result<(), BlockError> {
```

Comment:
Invalidation isn't only for blocks that aren't in the main-chain. We should totally be able to invalidate blocks in the main chain, which will result in invalidating all the blocks up to the tip. There are technical challenges there though for non-main-chain blocks.

And btw, side-chain doesn't mean what you think it does. A better name is "stale blocks". Side-chain refers to a blockchain network that has the ability to communicate to another blockchain through some form of bridging. For example, Polkadot uses parachains as side-chains. There's more details to this.

Reply:
> Invalidation isn't only for blocks that aren't in the main-chain. We should totally be able to invalidate blocks in the main chain, which will result in invalidating all the blocks up to the tip

Yes, I know, that's the other task that I'm working on.

> And btw, side-chain doesn't mean what you think it does. A better name is "stale blocks". 

Sorry, it was a slip of the hand, I know what side-chain means. I was going to use "off_chain_block" here or something similar. But "stale_block" also sounds good.